File: Javascript/Hashmap/M-187. Repeated DNA Sequences.ts
function findRepeatedDnaSequences(s: string): string[] {
  const map = new Map<string, number>();
  const result: string[] = [];

  for (let i = 0; i <= s.length - 10; i++) {
    const substr = s.substring(i, i + 10);
    const curCount = (map.get(substr) || 0) + 1;
    map.set(substr, curCount);
    if (curCount === 2) result.push(substr);
  }

  return result;
}

const testData = [
  { s: 'AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT', ans: ['AAAAACCCCC', 'CCCCCAAAAA'] },
  { s: 'AAAAAAAAAAAAA', ans: ['AAAAAAAAAA'] },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data['ans']) ===
      JSON.stringify(findRepeatedDnaSequences(data['s']))
  );
}

File: Hashmap/M-187. Repeated DNA Sequences/M-187. Repeated DNA Sequences.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> list[str]:
        map = {}
        result = []

        for i in range(len(s) - 9):
            substr = s[i : i + 10]
            count = (map.get(substr) or 0) + 1
            map[substr] = count

            if count == 2:
                result.append(substr)

        return result
